src/cluster.c

```diff
@@ -3525,14 +3525,165 @@ void clusterHandleManualFailover(void) {
+    if (nodeIsSlave(myself) && nodeIsMaster(node) && !nodeFailed(node)) {
+        int okslaves = clusterCountNonFailingSlaves(node);
+
+        /* A master is orphaned if it is serving a non-zero number of
+            * slots, have no working slaves, but used to have at least one
+            * slave, or failed over a master that used to have slaves. */
+        if (okslaves == 0 && node->numslots > 0 &&
+            node->flags & CLUSTER_NODE_MIGRATE_TO)
+        {
+            *orphaned_masters += 1;
+        }
+        if (okslaves > *max_slaves) *max_slaves = okslaves;
+        if (nodeIsSlave(myself) && myself->slaveof == node)
+            *this_slaves = okslaves;
+    }
+
+    /* If we are not receiving any data for more than half the cluster
+        * timeout, reconnect the link: maybe there is a connection
+        * issue even if the node is alive. */
+    mstime_t ping_delay = now - node->ping_sent;
+    mstime_t data_delay = now - node->data_received;
+    if (node->link && /* is connected */
+        now - node->link->ctime >
+        server.cluster_node_timeout && /* was not already reconnected */
+        node->ping_sent && /* we already sent a ping */
+        /* and we are waiting for the pong more than timeout/2 */
+        ping_delay > server.cluster_node_timeout/2 &&
+        /* and in such interval we are not seeing any traffic at all. */
+        data_delay > server.cluster_node_timeout/2)
+    {
+        /* Disconnect the link, it will be reconnected automatically. */
+        freeClusterLink(node->link);
+    }
+
+    /* If we have currently no active ping in this instance, and the
+        * received PONG is older than half the cluster timeout, send
+        * a new ping now, to ensure all the nodes are pinged without
+        * a too big delay. */
+    if (node->link &&
+        node->ping_sent == 0 &&
+        (now - node->pong_received) > server.cluster_node_timeout/2)
+    {
+        clusterSendPing(node->link, CLUSTERMSG_TYPE_PING);
+        return 0;
+    }
+
+    /* If we are a master and one of the slaves requested a manual
+        * failover, ping it continuously. */
+    if (server.cluster->mf_end &&
+        nodeIsMaster(myself) &&
+        server.cluster->mf_slave == node &&
+        node->link)
+    {
+        clusterSendPing(node->link, CLUSTERMSG_TYPE_PING);
+        return 0;
+    }
+
+    /* Check only if we have an active ping for this instance. */
+    if (node->ping_sent == 0) return 0;
+
+    /* Check if this node looks unreachable.
+        * Note that if we already received the PONG, then node->ping_sent
+        * is zero, so can't reach this code at all, so we don't risk of
+        * checking for a PONG delay if we didn't sent the PING.
+        *
+        * We also consider every incoming data as proof of liveness, since
+        * our cluster bus link is also used for data: under heavy data
+        * load pong delays are possible. */
```

Comment:
Some of your refactoring introduced some crazy spacing. Typically the * line up

```suggestion
    /* Check if this node looks unreachable.
     * Note that if we already received the PONG, then node->ping_sent
     * is zero, so can't reach this code at all, so we don't risk of
     * checking for a PONG delay if we didn't sent the PING.
     *
     * We also consider every incoming data as proof of liveness, since
     * our cluster bus link is also used for data: under heavy data
     * load pong delays are possible. */
```

Reply:
Ah thanks, i'll fix this when we get consensus on what to do `clusterNodeCronCheckFailingNodes`